File: Tree/94_BinaryTreeInorderTraversal.py
# coding: utf8


"""
    题目链接: https://leetcode.com/problems/binary-tree-inorder-traversal/description.
    题目描述:

    Given a binary tree, return the inorder traversal of its nodes' values.

    For example:
        Given binary tree [1,null,2,3],
           1
            \
             2
            /
           3
        return [1,3,2].

    Note: Recursive solution is trivial, could you do it iteratively?

"""


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ans = []
        # self.recursive_inorder(root, ans)
        # self.iterative_inorder(root, ans)
        self.morris_inorder(root, ans)

        return ans

    # 递归解法, 基本操作
    def recursive_inorder(self, root, ans):
        if not root:
            return

        self.recursive_inorder(root.left, ans)
        ans.append(root.val)
        self.recursive_inorder(root.right, ans)

    # 非递归解法, 与先序类似
    # 辅助栈
    def iterative_inorder(self, root, ans):
        stack = []
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            if stack:
                node = stack.pop()
                ans.append(node.val)
                root = node.right

    # 非递归解法
    # 不需要辅助栈, 充分利用叶子节点的空闲右指针, 指向当前节点的中序后继节点
    # 空间复杂度: O(1), 时间复杂度: O(n)
    def morris_inorder(self, root, ans):
        if not root:
            return

        cur = root
        while cur:
            if not cur.left:
                ans.append(cur.val)
                cur = cur.right
            else:
                pre = cur.left
                while pre.right and pre.right != cur:
                    pre = pre.right

                if pre.right == cur:
                    ans.append(cur.val)
                    pre.right = None
                    cur = cur.right
                else:
                    pre.right = cur
                    cur = cur.left

File: Tree/llb/94. Binary Tree Inorder Traversal.cpp
/*
问题描述：
Given a binary tree, return the inorder traversal of its nodes' values.

For example:
Given binary tree [1,null,2,3],
   1
    \
     2
    /
   3
return [1,3,2]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    
public:
    
    /*
    思路一:中序遍历,可以考虑使用递归的方式
    */
//     vector<int> inorderTraversal(TreeNode* root) {
       
//         vector<int> resVec;
//         inorder(root, resVec);
//         return resVec;
//     }
    
//     void inorder(TreeNode* root, vector<int> &vec) {
//         if (root == NULL) {
//             return;
//         }
        
//         inorder(root->left, vec);
//         vec.push_back(root->val);
//         inorder(root->right, vec);
//     }
    
    /*
    思路二：利用栈的机制
        由于中序遍历是左->根->右，所以利用栈可以。
        思路是从根节点开始，先将根节点压入栈，然后再将其所有左子结点压入栈，然后取出栈顶节点，保存节点值，
        再将当前指针移到其右子节点上，若存在右子节点，则在下次循环时又可将其所有左子结点压入栈中。
        这样就保证了访问顺序为左-根-右。
    */
    
//     vector<int> inorderTraversal(TreeNode* root) {
       
//         vector<int> resVec;
//         stack<TreeNode*> s;
//         TreeNode* p = root;
//         while (p || !s.empty()) {        //如果p不为null或者栈不为空就没有遍历完全
//             if (p) {
//                 s.push(p);
//                 p = p->left;
//             }
//             else {
//                 TreeNode *pTemp = s.top();
//                 s.pop();
//                 resVec.push_back(pTemp->val);
//                 p = pTemp->right;
//             }
//         }
        
//         return resVec;
//     }
    
    //写法三：就是先全部把左子树入栈，再一步一步往出弹
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> resVec;
        stack<TreeNode*> s;
        TreeNode* p = root;
        while (p || !s.empty()) {        //如果p不为null或者栈不为空就没有遍历完全
            while (p) {
                s.push(p);
                p = p->left;
            }
            
            p = s.top();
            s.pop();
            resVec.push_back(p->val);
            p = p->right;
        }
        
        return resVec;
    }
    
    
    
};
